File: java/有效的括号.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> string = new Stack<Character>(); //利用了栈的先入后出特性
        char a;
        for(int i = 0; i < s.length(); i++ ){
            a = s.charAt(i);
            if(string.empty()){
                string.push(a);
                continue;
            }
            if(a == ')' && string.peek() == '(')
                string.pop();
            else if(a == ']' && string.peek() == '[')
                string.pop();
            else if(a == '}' && string.peek() == '{')
                string.pop();
            else
                string.push(a);
        }
        if(string.empty())
            return true;
        return false;
    }
}

File: 初级算法/其他/有效的括号.cpp
class Solution {
public:
    bool isValid(string s) {
        
        string::iterator rightPos = s.begin();
        stack<char> temp;		//—π’ª∆•≈‰¿®∫≈
        
        bool result = true;
        
        if(s.size() % 2 == 1) 
            return false;
        if(*rightPos == ')' || *rightPos == '}' || *rightPos == ']')
            return false;
        
        while(rightPos != s.end()){
            if(*rightPos == '(' || *rightPos == '{' || *rightPos == '[')
                temp.push(*rightPos);
            else{
                char tempChar = temp.top();
                temp.pop();
                if(tempChar == '{' && *rightPos != '}'){
                    result = false;
                    break;
                }
                if(tempChar == '(' && *rightPos != ')'){
                    result = false;
                    break;
                }
                if(tempChar == '[' && *rightPos != ']'){
                    result = false;
                    break;
                }
            }
            ++ rightPos;
        }
        if(temp.empty())
            return result;
        else
            return false;
    }
};